### PR TITLE
Add published date feature

### PR DESCRIPTION
**Pull request by matthieu-rolland**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | add a `published_date` field on product creation/edition form that can be used with newproduct module to set a date of publication that is different from the date creation
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | 
| UI Tests          | 
| Fixed issue or discussion?     | 
| Related PRs       | 
| Sponsor company   | PrestaShop SA

### CODE REVIEW

**Review by Hlavtox — CHANGES_REQUESTED** [review]
Why extra column? Just add a field that controls date_add and solved. That's what the column is for. :-)

**Comment by matthieu-rolland** [comment]
> Why extra column? Just add a field that controls date_add and solved. That's what the column is for. :-)

That's how it was specified to me...  but I think it's better to have a dedicated column, `date_add` is a column we have on many tables, and that can't be modified through the BO, it's supposed to keep the real date of creation; I don't want to make an exception in the way it behaves just for Product.

**Comment by Hlavtox** [comment]
@matthieu-rolland Another column is just more data to manage with which almost never change and will me 99% cases the same as date_add.

Also, it could create data inconsistencies when we have to keep two columns in sync.

And the biggest issue is that it will break so many things in the ecosystem that depend on date_add to get new product logic. Filtering modules, tags on products, import modules, core listings, everything.

I was 100% for the change, but if it would mean this, no way.

Ping @ShaiMagal

**Review by PrestaEdit — APPROVED** [review]

**Comment by PrestaEdit** [comment]
Date creation is not the same. Also, published date could be changed on every update of the active field, maybe.

**Comment by ShaiMagal** [comment]
@Hlavtox @matthieu-rolland 
Guys, my opinion I will little bit mix it :-D
40% for Hlavtox and 40% for Matthieu and 20% for my opinion

 - we should keep `date_add` with same behaviour (same connections for other things like other modules etc), creating new table and connecting "everything again" is not good idea I think. There are connections with 3rd party modules for example.
 
 - if we want to allow eshop owner to change creation date of product, we can allow them to edit `date_add`, but in every case we **NEED** to store database column with **REAL** date of creation. What do you think about `date_pub` (shortcut for "date published")
 
Result of my thoughts:
  - edit this PR to edit `date_add` (don't add new column now)
  - add new PR with new DB column `date_pub` (where will be filled REAL date creating of product, without possibility to change and eshop owner can change `date_add`)
  - don't use `published_date`, stay "semantic", so `date_*`
  - both PR to develop branch of course
  
  We need to know real date of product creation while maintaining current functionality and continuity.

**WDYT?**

**Comment by kpodemski** [comment]
I think it's the change not worth implementing, sorry

there are some modules that allow to force some products to be displayed as new by modifying date add field, and it's a good workaround

introducing this new column requires changing all product queries so that "new" label is correctly displayed

is it really worth developers time with only a few of them working on the core? 😱 

and from what I see it's not only about handling "new" label, it's also about allowing people to choose when the product should be visible

it is first and foremost bc break, secondly it won't work in most of the modules because we don't have cache invalidation mechanism that could handle it

sorry to sound harsh but I'm pointing obvious things that should be taken into consideration sooner

**Comment by matthieu-rolland** [comment]
I will let you guys discuss this with @MatShir, I just want to clarify some misunderstanding on the goal of this feature.

> 
> and from what I see it's not only about handling "new" label, it's also about allowing people to choose when the product should be visible
> 


Not really, this field would have no impact on product visibility, from the specs given to me, (you can tell me if I misunderstood @MatShir ), it's meant to be used by the newproduct module, which will have a pull request too.

Let's say a product was out of stock for a long time, and it's reappearing, the merchant wants it to appear in the "new module" slider. 

Maybe the field is badly named though, `date_published` is misleading. Or maybe the field should be added by the module itself and not on the core itself, it would make more sense ?

>
> there are some modules that allow to force some products to be displayed as new by modifying date add field, and it's a good workaround
>
>

Imho modifying the `date_add` field is not a good workaround, the content of this field should be trustworthy and never change, it's the date of product creation in the database. We have the same field on many tables, its intent is very clear.

File: src/PrestaShopBundle/Form/Admin/Sell/Product/Options/OptionsType.php [hunk 1]
```diff
@@ -42,6 +43,14 @@ class OptionsType extends TranslatorAwareType
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
         $builder
+            ->add('published_date', DatePickerType::class, [
+                'required' => false,
+                'label' => $this->trans('Date of publication', 'Admin.Catalog.Feature'),
+                'label_subtitle' => $this->trans('When do you want your product to appear?', 'Admin.Catalog.Feature'),
```
Comment (kpodemski) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/Options/OptionsType.php:49]: @matthieu-rolland this suggests it's a feature to plan the publication of the product

Comment (matthieu-rolland) [reply src/PrestaShopBundle/Form/Admin/Sell/Product/Options/OptionsType.php:49]: true, I knew this was bad, I copy pasted `Where do you want your product to appear?` and replaced `appear` with `published`, I know it's wrong, I'm just waiting for proper label with @MatShir 

that's why this PR is in draft, I just opened it to check the CI... 

We discussed with @MatShir about this "published" naming and havent found a better naming yet, it's work in progress

**Comment by PrestaEdit** [comment]
Could, maybe, have one ask ?

I could read more than one times on this thread :
- That's how it was specified to me...
- I was 100% for the change
- from the specs given to me

And... nothing related : no issues, no ADR, no discussions, ...
So : where are theses specs and how are they born ? 🤔

**Comment by matthieu-rolland** [comment]
https://github.com/orgs/PrestaShop/projects/7/views/73?pane=issue&itemId=50332495

**Comment by PrestaEdit** [comment]
Error 404 for non PrestaShop SA or assimilated.

**Comment by Hlavtox** [comment]
@matthieu-rolland @MatShir It's not just the ps_newproducts module, it's the whole ecosystem.

What about faceted search? What about listing of new products on `/new-products` page? Export modules that depend on some labels? Everything will be broken.

Why we can't just go for a simple solution to add one single field that would change `date_add`?

BTW - Where is some discussion with tech council or product council? Idea created 3 days ago and consulted with no one, if somebody asked before, you would not have to loose time with a feature that maybe won't get accepted.

**Comment by matthieu-rolland** [comment]
> Why we can't just go for a simple solution to add one single field that would change `date_add`?

Can you develop ?

**Comment by matthieu-rolland** [comment]
> * edit this PR to edit `date_add` (don't add new column now)
> * add new PR with new DB column `date_pub` (where will be filled REAL date creating of product, without possibility to change and eshop owner can change `date_add`)

The problem is that we have this `date_add` column on a lot of tables, and it feels strange to have a different behavior for this column only on the product table.  The intent of this field is to have the date of creation of the item on the database, it should never be modified.

The `date_add` field is filled directly from the object model if it exists at item creation (in the `add` method)

**Comment by kpodemski** [comment]
The idea could be to have a checkbox to force product to be displayed on the new products page, in the module, have a label, etc. Same as with on_sale flag.

**Comment by matthieu-rolland** [comment]
> The idea could be to have a checkbox to force product to be displayed on the new products page, in the module, have a label, etc. Same as with on_sale flag.

to be discussed with @MatShir but it sounds very reasonable

**Comment by Hlavtox** [comment]
@kpodemski It must be automatic, why would you introduce manual work for you? Have fun setting the flag manually on 200 products you just added to your catalog, then setting it back. 🤦‍♂️

**Comment by MatShir** [comment]
Here the issue with the idea and specs https://github.com/PrestaShop/PrestaShop/issues/35447 (never to late :p)

**Comment by matthieu-rolland** [comment]
> @kpodemski It must be automatic, why would you introduce manual work for you? Have fun setting the flag manually on 200 products you just added to your catalog, then setting it back. 🤦‍♂️

it could be both, automatic, and possibility to fine tune with a flag

**Comment by Hlavtox** [comment]
It worked fine for 15 years, why we can't change date_add when it brings no dangers? You can just do $objectModel->add(false) and you can set your own date of creation.

It would be finished in 10 minutes and it would be a quick win.

**Comment by kpodemski** [comment]
I see we have three fighters :)

1. Manual work for merchants, adding a flag `is_new`.
2. Painful change, adding `publication_date` field.
3. Not-ideal solution with refreshing `date_add`, a field that is about adding the product (object) to the shop.

So, if nobody comes up with a better solution, I would choose the third fighter as it is something that would leave the ecosystem intact while resolving the problem. No need to modify any front office queries, showing products from the last X days would work the same, etc.

I see arguments that `date_add` shouldn't be touched, well... I'm not sure if it's dangerous to be honest :)

**Comment by matks** [comment]
Hello everyone, I want to quote the [Code of Conduct](https://github.com/PrestaShop/PrestaShop/blob/31f93eef72b04b6930aee6dfb2cec4ded4909ca9/CODE_OF_CONDUCT.md):

> Examples of behavior that contributes to creating a positive environment include:

> Using welcoming and inclusive language
Being respectful of differing viewpoints and experiences
Gracefully accepting constructive criticism
Focusing on what is best for the community
Showing empathy towards other community members

IMO some of the messages here do NOT contribute to create a positive environment.

Discussing pros and cons of an implementation, suggesting ideas, pointing out issues and weaknesses are all good, but I believe it can be done with the right tone of voice.

Some messages in this discussion are 100% OK, but some are not. I think we can all focus on delivering a good implementation and set aside the un-needed hasty judgment.

**Comment by matks** [comment]
> Could, maybe, have one ask ?
> 
> I could read more than one times on this thread :
> 
> * That's how it was specified to me...
> * I was 100% for the change
> * from the specs given to me
> 
> And... nothing related : no issues, no ADR, no discussions, ... So : where are theses specs and how are they born ? 🤔

@PrestaEdit People in the PrestaShop project organization do have private communication channels. For example we have multiple private slack channels. The list is available here https://www.prestashop-project.org/slack/channels/#active-channels . These channels are not available to people outside of the project organization.

**Comment by matthieu-rolland** [comment]
FYI this subject has been blocked and will be studied again by the product and tech team in the future, I'm sure you'll be in the loop @Hlavtox  👍  Meanwhile I'll take on other subjects

**Comment by matthieu-rolland** [comment]
FYI this subject has been blocked and will be studied again by the product and tech team in the future, I'm sure you'll be in the loop @Hlavtox  👍  Meanwhile I'll take on other subjects

**Comment by PrestaEdit** [comment]
> > Could, maybe, have one ask ?
> > I could read more than one times on this thread :
> > 
> > * That's how it was specified to me...
> > * I was 100% for the change
> > * from the specs given to me
> > 
> > And... nothing related : no issues, no ADR, no discussions, ... So : where are theses specs and how are they born ? 🤔
> 
> @PrestaEdit People in the PrestaShop project organization do have private communication channels. For example we have multiple private slack channels. The list is available here https://www.prestashop-project.org/slack/channels/#active-channels . These channels are not available to people outside of the project organization.

It's why I ask.

Private discussion are and was already be done on this project, as far I remember. It's not an open-wide universe, PrestaShop.

But as far I know, when something is published on the public side of the project, it's not more private.

PrestaShop Project organization ask us to have an issue even for a minor change, and to link it to the Pull Request as possible.
There, there is nothing related and finally we can could have something published (later, of course) about it.

And as I can read : there is only a hand of people that could give their opinion on this change (and it's seems not to have the whole Product Council implicated).

Or you keep the "open" side "open" as needed, using discussion (even if it's not taking feedback on them, obviously) or you close the "open" side of this project, as now.

**Comment by matks** [comment]
> (even if it's not taking feedback on them, obviously)

Ah, I was waiting to see when you would seize the opportunity to slip a little critic.

> (later, of course)

And here comes the sarcasm too.

I had prepared a 15-lines long answer to your message but seeing as you mix critics and sarcasm into your message, I change my mind and simply don't want to answer 🤷‍♂️ . I'll be happy answering this question if it comes without the critics, without the sarcasm.

**Comment by PrestaEdit** [comment]
Why it's always the same, there ? When it's not as expected/in your mind, it's not positive and it's quickly sarcasm or critiscim ?

Really, I don't understand the move of this project since years, now.

Isn't it true ? Really, could you say that it's not true ?

I'm not against this change or another, I try to understand why this kind of subject are handled on the open source project that be called PrestaShop. Simply ;-)

**Comment by matthieu-rolland** [comment]
> PrestaShop Project organization ask us to have an issue even for a minor change, and to link it to the Pull Request as possible. There, there is nothing related and finally we can could have something published (later, of course) about it.

because this PR was not ready for review, it was just a draft, the `draft` feature is supposed to mean that the PR is not ready for review (either the code or the description). I opened it yesterday evening because I wanted to see what the CI's result would be instead of running it on my computer...

I was really not expecting this DRAFT PR to gather so much attention, next time I'll open my draft PRs on a private PrestaShop fork to avoid any kind of drama.

**Comment by PrestaEdit** [comment]
:)

**Pull request closed without merging**